File: 11/ghost/p.py
from sys import stdin, stdout

"""
Trie implementation from https://www.geeksforgeeks.org/trie-insert-and-search/
"""
class TrieNode:
     
    # Trie node class
    def __init__(self):
        self.children = [None]*26
 
        # isEndOfWord is True if node represent the end of the word
        self.isEndOfWord = False
 
class Trie:
     
    # Trie data structure class
    def __init__(self):
        self.root = self.getNode()
 
    def getNode(self):
     
        # Returns new trie node (initialized to NULLs)
        return TrieNode()
 
    def _charToIndex(self,ch):
         
        # private helper function
        # Converts key current character into index
        # use only 'a' through 'z' and lower case
         
        return ord(ch)-ord('a')
 
 
    def insert(self,key):
         
        # If not present, inserts key into trie
        # If the key is prefix of trie node,
        # just marks leaf node
        pCrawl = self.root
        length = len(key)
        for level in range(length):
            index = self._charToIndex(key[level])
 
            # if current character is not present
            if not pCrawl.children[index]:
                pCrawl.children[index] = self.getNode()
            pCrawl = pCrawl.children[index]
 
        # mark last node as leaf
        pCrawl.isEndOfWord = True
 
    def search(self, key):
         
        # Search key in the trie
        # Returns true if key presents
        # in trie, else false
        pCrawl = self.root
        length = len(key)
        for level in range(length):
            index = self._charToIndex(key[level])
            if not pCrawl.children[index]:
                return False
            pCrawl = pCrawl.children[index]
 
        return pCrawl.isEndOfWord



def firstHaveStrategyToWin(pos):
    if pos.isEndOfWord: return True

    for i in range(26):
        nextPos = pos.children[i]
        if nextPos != None:
            if not firstHaveStrategyToWin(nextPos):
                return True
    return False

def firstHaveStrategyToLoose(pos):
    if pos.isEndOfWord: return False

    for i in range(26):
        nextPos = pos.children[i]
        if nextPos != None:
            if not firstHaveStrategyToLoose(nextPos):
                return True
    return False

def main():
    t = int(stdin.readline())
    for testcase in range(t):
        n,w = stdin.readline().split()
        n = int(n)
        w = int(w)

        words = []
        for i in range(w):
            temp = stdin.readline()
            words.append(temp[:-1].lower())

        t = Trie()
        for w in words:
            t.insert(w)

        leaHaveWinStrategy = firstHaveStrategyToWin(t.root)
        leaHaveLooseStrategy = firstHaveStrategyToLoose(t.root)
        
        #Scenario 1 (Lea begin, winner begin)
        scenario1 = False
        if leaHaveWinStrategy:
            scenario1 = True
        else:
            if leaHaveLooseStrategy:
                scenario1 = n%2==0
            else:
                scenario1 = False
        
        #Scenario 2 (Lea begin, looser begin)
        scenario2 = False
        if not leaHaveWinStrategy:
            scenario2 = False
        else:
            if not leaHaveLooseStrategy:
                scenario2 = not n%2==0
            else:
                scenario2 = True
        
        #Scenario 3 and 4 are deduced from the symetry of the game
        scenario3 = not scenario1
        scenario4 = not scenario2

        print("Case #"+str(testcase+1)+":")
        if scenario1: print("victory")
        else: print("defeat")
        if scenario2: print("victory")
        else: print("defeat")
        if scenario3: print("victory")
        else: print("defeat")
        if scenario4: print("victory")
        else: print("defeat")

        stdin.readline()

if __name__ == "__main__":
    main()

File: 11/ghost/p.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <functional>
#include <cmath>
#include <array>
#include <vector>
#include <list>
#include <deque>
#include <queue>
#include <stack>
#include <map>
#include <set>
#include <unordered_map>
#include <bitset>
#include <chrono>
#include <climits>
#include <cfloat>

using namespace std;

#define SI(x) scanf("%d",&x)
#define SII(x,y) scanf("%d %d",&x,&y)
#define SIII(x,y,z) scanf("%d %d %d",&x,&y,&z)
#define SIIII(x,y,z,w) scanf("%d %d %d %d",&x,&y,&z,&w)
#define SL(x) scanf("%lld",&x)
#define SD(x) scanf("%lf",&x)
#define SC(x) scanf("%c",&x)
#define SS(x) scanf("%s",x)
#define FOR(i, s, k) for(int i=s; i<k; i++)
#define REP(i, n) FOR(i, 0, n)
#define INF INT_MAX
#define EPS 1e-9
#define PI acos(-1)

typedef long long int lint;
typedef unsigned long long int ulint;
typedef pair<int, int> pii;
typedef pair<lint, lint> pll;
typedef pair<double, double> pdd;
typedef vector<int> vi;
typedef vector<lint> vl;
typedef vector<double> vd;
typedef vector<bool> vb;
typedef list<int> li;
typedef list<double> ld;
typedef list<bool> lb;
typedef vector<vector<int>> vvi;
typedef vector<vector<lint>> vvl;
typedef vector<vector<double>> vvd;
typedef vector<pair<int, int>> vii;
typedef list<pair<int, int>> lii;
typedef vector<list<int>> vli;
typedef vector<list<pair<int, int>>> vlii;
typedef vector<list<pair<double,int>>> vldi;


/*****************************************************************************
Trie implementation from https://www.geeksforgeeks.org/trie-insert-and-search/
*****************************************************************************/
const int ALPHABET_SIZE = 26;
 
// trie node
struct TrieNode
{
    struct TrieNode *children[ALPHABET_SIZE];
 
    // isEndOfWord is true if the node represents
    // end of a word
    bool isEndOfWord;
};
 
// Returns new trie node (initialized to NULLs)
struct TrieNode *getNode(void)
{
    struct TrieNode *pNode =  new TrieNode;
 
    pNode->isEndOfWord = false;
 
    for (int i = 0; i < ALPHABET_SIZE; i++)
        pNode->children[i] = NULL;
 
    return pNode;
}
 
// If not present, inserts key into trie
// If the key is prefix of trie node, just
// marks leaf node
void insert(struct TrieNode *root, string key)
{
    struct TrieNode *pCrawl = root;
 
    for (int i = 0; i < key.length(); i++)
    {
        int index = key[i] - 'a';
        if (!pCrawl->children[index])
            pCrawl->children[index] = getNode();
 
        pCrawl = pCrawl->children[index];
    }
 
    // mark last node as leaf
    pCrawl->isEndOfWord = true;
}
 
// Returns true if key presents in trie, else
// false
bool search(struct TrieNode *root, string key)
{
    struct TrieNode *pCrawl = root;
 
    for (int i = 0; i < key.length(); i++)
    {
        int index = key[i] - 'a';
        if (!pCrawl->children[index])
            return false;
 
        pCrawl = pCrawl->children[index];
    }
 
    return (pCrawl->isEndOfWord);
}


bool firstHaveStrategyToWin(TrieNode* pos){
    if(pos->isEndOfWord) return true;

    REP(i,26){
        TrieNode* nextPos = pos->children[i];
        if(nextPos != NULL){
            if(!firstHaveStrategyToWin(nextPos)){
                return true;
            }
        }
    }
    return false;
}

bool firstHaveStrategyToLoose(TrieNode* pos){
    if(pos->isEndOfWord) return false;

    REP(i,26){
        TrieNode* nextPos = pos->children[i];
        if(nextPos != NULL){
            if(!firstHaveStrategyToLoose(nextPos)){
                return true;
            }
        }
    }
    return false;
}



int main(){

    int t; SI(t); FOR(testcase, 1, t+1){
		int n,w; SII(n,w);
        vector<string> words(w);
        REP(i,w) cin>>words[i];

        TrieNode* root = getNode();
        REP(i,w) insert(root, words[i]);

        bool leaHaveWinStrategy = firstHaveStrategyToWin(root);
        bool leaHaveLooseStrategy = firstHaveStrategyToLoose(root);

        //Scenario 1 (Lea begin, winner begin)
        bool scenario1;
        if(leaHaveWinStrategy){
            scenario1 = true;
        }else{
            if(leaHaveLooseStrategy){
                scenario1 = n%2==0;
            }else{
                scenario1 = false;
            }
        }

        //Scenario 2 (Lea begin, looser begin)
        bool scenario2;
        if(!leaHaveWinStrategy){
            scenario2 = false;
        }else{
            if(!leaHaveLooseStrategy){
                scenario2 = !(n%2==0);
            }else{
                scenario2 = true;
            }
        }

        //Scenario 3 and 4 are deduced from the symetry of the game
        bool scenario3 = !scenario1;
        bool scenario4 = !scenario2;

		printf("Case #%d:\n", testcase);
        if(scenario1) printf("victory\n");
        else printf("defeat\n");
        if(scenario2) printf("victory\n");
        else printf("defeat\n");
        if(scenario3) printf("victory\n");
        else printf("defeat\n");
        if(scenario4) printf("victory\n");
        else printf("defeat\n");
    }

    return 0;
}
